### PR TITLE
Refactor of SuggestedSearch logic. [wip]

### PR DESCRIPTION
**Pull request by timabell**
Did this as part of trying to understand existing code.

Needs regression tests before merging.

### Context
Code was hard to follow

### Changes proposed in this pull request
Should be a pure refactor, mistakes not withstanding.

### CODE REVIEW

**Comment by timabell** [comment]
[Fix link to suggested courses if no results](https://trello.com/c/GMXhEE4M/381-fix-link-to-suggested-courses-if-no-results)

File: src/ViewComponents/SuggestedSearch.cs [hunk 1]
```diff
@@ -22,39 +22,46 @@ public SuggestedSearch(ISearchAndCompareApi api)
...
-                result.SuggestedSearches.Add(salaryResult);
+                var anyFundingFilter = new List<ResultsFilter> { RemoveFundingFilter(originalViewModel) };
+                salaryResults = GetViewModelsWithCounts(anyFundingFilter, originalTotal, anyFundingFilter.Count());
             }
-            else {
-                result.SuggestedSearches = GetResults(resultsFilters, originalTotal, maxResult);
+            var result = new SuggestedSearchesViewModel { OriginalResults = originalViewModel, HasSalary = hasSalary };
+            var numberOfSearchesToBuild = maxResult;
+            // always make the last entry the all-funding one if salary was set
+            if (salaryResults.Any()) {
```
Comment (defong) [inline src/ViewComponents/SuggestedSearch.cs]: This should be nested inside of `if (hasSalary) {`

File: src/ViewComponents/SuggestedSearch.cs [hunk 2]
```diff
@@ -22,39 +22,46 @@ public SuggestedSearch(ISearchAndCompareApi api)
...
             }
-
-
-            return Task.FromResult<IViewComponentResult>(View(result));
+            result.SuggestedSearches = GetViewModelsWithCounts(filters, originalTotal, numberOfSearchesToBuild);
+            return result;
         }
 
-        private ResultsFilter GetSuggestedSearchesResultsFilterForAnyFunding(ResultsViewModel original)
+        private ResultsFilter RemoveFundingFilter(ResultsViewModel original)
```
Comment (defong) [inline src/ViewComponents/SuggestedSearch.cs]: Misleading? 

`Remove funding` vs `Any funding`

File: src/ViewComponents/SuggestedSearch.cs [hunk 3]
```diff
@@ -22,39 +22,46 @@ public SuggestedSearch(ISearchAndCompareApi api)
...
+                // put usual suggestions first, but make room for salary suggestions within max
+                result.SuggestedSearches = GetViewModelsWithCounts(filters, originalTotal, maxResult - salaryResults.Count() );
+                result.SuggestedSearches.Add(salaryResults.First());
+            } else {
+                result.SuggestedSearches = GetViewModelsWithCounts(filters, originalTotal, maxResult);
             }
-
-
-            return Task.FromResult<IViewComponentResult>(View(result));
+            result.SuggestedSearches = GetViewModelsWithCounts(filters, originalTotal, numberOfSearchesToBuild);
```
Comment (defong) [inline src/ViewComponents/SuggestedSearch.cs]: Is this a shorthand else or is it intentional? confusing

**Comment by timabell** [comment]
Extract a testable service

**Comment by timabell** [comment]
[refactor & add tests to suggested search code](https://trello.com/c/TXLBOoFQ/440-refactor-add-tests-to-suggested-search-code)
